src/pb_stub.cc

```diff
@@ -848,52 +854,69 @@ Stub::GetOrCreateInstance()
 }
 
 void
-Stub::LaunchLogRequestThread()
+Stub::LaunchStubToParentQueueMonitor()
 {
-  log_thread_ = true;
-  log_monitor_ = std::thread(&Stub::ServiceLogRequests, this);
+  stub_to_parent_thread_ = true;
+  stub_to_parent_queue_monitor_ =
+      std::thread(&Stub::ServiceStubToParentRequests, this);
   Logger::GetOrCreateInstance()->SetBackendLoggingActive(true);
 }
 
 void
-Stub::TerminateLogRequestThread()
+Stub::TerminateStubToParentQueueMonitor()
 {
   Logger::GetOrCreateInstance()->SetBackendLoggingActive(false);
   {
-    std::lock_guard<std::mutex> guard{log_message_mutex_};
+    std::lock_guard<std::mutex> guard{stub_to_parent_message_mu_};
     log_request_buffer_.push(DUMMY_MESSAGE);
+    bls_response_cleanup_buffer_.push(DUMMY_MESSAGE);
   }
-  log_message_cv_.notify_one();
-  log_monitor_.join();
+  stub_to_parent_message_cv_.notify_one();
+  stub_to_parent_queue_monitor_.join();
 }
 
 void
 Stub::EnqueueLogRequest(std::unique_ptr<PbLog>& log_ptr)
 {
   {
-    std::lock_guard<std::mutex> guard{log_message_mutex_};
+    std::lock_guard<std::mutex> guard{stub_to_parent_message_mu_};
     log_request_buffer_.push(std::move(log_ptr));
   }
-  log_message_cv_.notify_one();
+  stub_to_parent_message_cv_.notify_one();
 }
 
 void
-Stub::ServiceLogRequests()
+Stub::ServiceStubToParentRequests()
 {
-  while (log_thread_) {
-    std::unique_lock<std::mutex> guard{log_message_mutex_};
-    while (log_request_buffer_.empty()) {
-      log_message_cv_.wait(guard);
+  while (stub_to_parent_thread_) {
+    std::unique_lock<std::mutex> guard{stub_to_parent_message_mu_};
+    while (log_request_buffer_.empty() &&
+           bls_response_cleanup_buffer_.empty()) {
+      stub_to_parent_message_cv_.wait(guard);
     }
-    // On exit, will send messages until
-    // DUMMY_MESSAGE is reached
-    std::unique_ptr<PbLog> log_request = std::move(log_request_buffer_.front());
-    if (log_request == DUMMY_MESSAGE) {
-      log_request_buffer_.pop();
-      break;
-    } else {
-      log_request_buffer_.pop();
-      SendLogMessage(log_request);
+    if (!log_request_buffer_.empty()) {
+      // On exit, will send messages until
+      // DUMMY_MESSAGE is reached
+      std::unique_ptr<PbLog> log_request =
+          std::move(log_request_buffer_.front());
+      if (log_request == DUMMY_MESSAGE) {
+        log_request_buffer_.pop();
+        break;
+      } else {
+        log_request_buffer_.pop();
+        SendLogMessage(log_request);
+      }
+    }
+    if (!bls_response_cleanup_buffer_.empty()) {
+      void* id = std::move(bls_response_cleanup_buffer_.front());
+      if (id == DUMMY_MESSAGE) {
+        bls_response_cleanup_buffer_.pop();
+        break;
+      } else {
+        bls_response_cleanup_buffer_.pop();
+        SendCleanupId(id);
+        response_iterator_map_.erase(id);
+      }
```

Comment:
I think we should merge these queues into a single queue and have `command` to distinguish between them. This doesn't need to be fixed right now but we probably need to fix it later.

Reply:
Sounds good! Let me make this change in another PR.